### ISSUE TITLE
conditional dependencies

### ISSUE THREAD

**Issue by phi-gamma** [sagiegurari/cargo-make#304]
**Features Description**

Provide a mechanism for choosing or overriding a dependency based
on profiles.

**Describe the solution you'd like**

Ideally, parameters from an “env” would be expanded in a
“dependencies” expression:

    [env.debug]
    build_mode   = "build"

    [env.release]
    build_mode   = "build-release"

    [tasks.foo]
    dependencies = [ "${build_mode}" ] ## ← error, unexpanded

This would be the most straightforward option and also consistent
with how env expansion works elsewhere. 

Another option could be extending an existing task conditionally:

    [tasks.foo]
    dependencies = [ "build" ]

    ## error: redefinition of table `tasks.foo` for key `tasks.foo`
    [tasks.foo]
    condition    = { profiles = [ "release" ] }
    dependencies = [ "build-release" ]

Those are just suggestions though based on what I intuitively
tried.

**Code Sample**

This is a distilled version of a real world makefile that
compiles and links a C program which depends on a cdylib built by
cargo. The profiles “debug” and “release” result in different
sets of flags being passed to the compiler and linker, and pick
the correct paths to build artifacts generated by cargo.

The entry point is task “bar” which needs to depend on “build”
under the debug profile but on “build-release” for the release
one. This is in essence the problem to be solved by the feature
request.

    [env]
    cflags          = "-Wall --std=gnu99"
    ldflags         = "-lfoo"
    
    bar             = "bar"
    bar_src         = "src/bar/${bar}.c"
    bar_obj         = "${bar}.o"
    
    [env.debug]
    target_dir      = "target/debug"
    cflags          = "${cflags} -Wfatal-errors -Werror -Og -ggdb -DDEBUG"
    
    [env.release]
    target_dir      = "target/release"
    cflags          = "${cflags} -O3 -DNDEBUG"
    
    [tasks.compile-bar]
    command         = "gcc"
    args            = [ "@@split(cflags, )"
                      , "-Isrc/foo
                      , "-o", "${target_dir}/${bar_obj}"
                      , "-c", "${bar_src}"
                      ]
    ## XXX This should be “build-release” when the release profile is active.
    dependencies    = [ "build" ]
    
    [tasks.link-bar]
    command         = "gcc"
    args            = [ "@@split(ldflags, )"
                      , "-L${target_dir}"
                      , "-o", "${target_dir}/${bar}"
                      , "${target_dir}/${bar_obj}"
                      ]
    dependencies    = [ "compile-bar" ]
    
    [tasks.bar]
    dependencies    = [ "link-bar" ]

**Comment by sagiegurari** [comment]
nice. lets see if existing capabilities already provide a solution for you or something else is needed.
let's take the build task as an example:

```toml
[env]
CARGO_MAKE_CARGO_ALL_FEATURES = "--all-features"
CARGO_MAKE_CARGO_BUILD_TEST_FLAGS = "${CARGO_MAKE_CARGO_ALL_FEATURES}"

[tasks.build]
description = "Runs the rust compiler."
category = "Build"
command = "cargo"
args = ["build", "@@split(CARGO_MAKE_CARGO_BUILD_TEST_FLAGS, )"]
```

I took these parts from the internal makefile.toml.
anyhow, as you can see, the ```cargo build``` command accepts additional optional params.
by default those params are **--all-features** but it can also be empty or you can define **--release** as the value or multiple values and so on...

I also have the following:

```toml
[env]
CARGO_MAKE_CARGO_PROFILE = { source = "${CARGO_MAKE_PROFILE}", mapping = {"development" = "dev", "ci" = "test" } }
```

which  provides you of a cargo profile value based on existing cargo-make profiles.
so you could do

```toml
[env]
CARGO_MAKE_CARGO_BUILD_TEST_FLAGS = "--${CARGO_MAKE_CARGO_PROFILE}"
```

which will result in --release or --dev provided in the cargo build command based if you ran cargo-make with development profile or release profile.

Bottom line, it means that the same task definition can do many different things and you don't have to use multiple task definitions for same operation with different arguments.

This is existing capability, is it providing you a simple solution?

**Comment by phi-gamma** [comment]
Thanks for the detailed write-up. I added a line tweaking ``CARGO_MAKE_CARGO_BUILD_TEST_FLAGS`` to the config and it works as desired.

> This is existing capability, is it providing you a simple solution?

Absolutely! Closing this issue as redundant.
